### PR TITLE
feat(oas): merge example rules

### PR DESCRIPTION
**Pull request by P0lip**
This is a follow-up PR.
See https://github.com/stoplightio/spectral/pull/1284#issuecomment-678874684 for the reasoning.

tl;dr we no longer need such granular rules, as `given` can be an array of paths now.

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/rulesets/oas/__tests__/oas2-valid-schema-example.ts [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+      expect(results).toEqual([
+        expect.objectContaining({
+          code: 'oas2-valid-schema-example',
+          message: '`example` property should have required property `url`',
+          severity: DiagnosticSeverity.Error,
+        }),
+      ]);
+    });
+
+    test('will pass for valid parents examples which contain invalid child examples', async () => {
```
Comment (nulltoken) [inline src/rulesets/oas/__tests__/oas2-valid-schema-example.ts:130]:
```suggestion
    test('will fail for valid parents examples which contain invalid child examples', async () => {
```

File: src/rulesets/oas/__tests__/oas2-valid-schema-example.ts [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+        expect.objectContaining({
+          severity: DiagnosticSeverity.Error,
+          code: 'oas2-valid-schema-example',
+          message: '`self` property type should be array',
+          path: ['definitions', 'halRoot', 'example', '_links', 'self'],
+        }),
+      ]);
+    });
+
+    test('will pass for valid parents examples which contain invalid child examples', async () => {
```
Comment (nulltoken) [inline src/rulesets/oas/__tests__/oas2-valid-schema-example.ts:435]:
```suggestion
    test('will fail for valid parents examples which contain invalid child examples', async () => {
```

File: src/rulesets/oas/__tests__/oas3-valid-schema-example.ts [hunk 3]
```diff
@@ -15,228 +15,801 @@ describe('oas3-valid-schema-example', () => {
...
+              type: 'string',
+              example: 'doggie',
+            },
+          },
+        ],
+      });
+      expect(results).toHaveLength(0);
+    });
+
+    test('will pass for valid parents examples which contain invalid child examples', async () => {
```
Comment (nulltoken) [inline src/rulesets/oas/__tests__/oas3-valid-schema-example.ts:260]:
```suggestion
    test('will fail for valid parents examples which contain invalid child examples', async () => {
```

File: src/rulesets/oas/__tests__/oas3-valid-schema-example.ts [hunk 4]
```diff
@@ -15,228 +15,801 @@ describe('oas3-valid-schema-example', () => {
...
+      expect(results).toEqual([
+        expect.objectContaining({
+          severity: DiagnosticSeverity.Error,
+          code: 'oas3-valid-schema-example',
+          message: '`example` property type should be string',
+        }),
+      ]);
+    });
+
+    test('will pass for valid parents examples which contain invalid child examples', async () => {
```
Comment (nulltoken) [inline src/rulesets/oas/__tests__/oas3-valid-schema-example.ts:562]:
```suggestion
    test('will fail for valid parents examples which contain invalid child examples', async () => {
```

**Comment by nulltoken** [comment]
@P0lip Great job! One thought though. There are a lot of code duplication among the tests, and some tests exist in some file and not in others (`format` related ones IIRC for instance). 

What would you think of extracting the valid/invalid schemas and the matching props of the results we'd like to check in one global collection, and then "inject" the schemas in proper locations of base documents in each test? We may have to tweak the expectations regarding the resulting paths, but it's possible this would greatly reduce the size of the diff and (maybe) make those tests a bit easier to maintain and reason about?

**Comment by P0lip** [comment]
Agree re duplication. It bothers me too.

> What would you think of extracting the valid/invalid schemas and the matching props of the results we'd like to check in one global collection, and then "inject" the schemas in proper locations of base documents in each test?

Hm, not sure I'm following. What do you exactly suggest?

**Comment by philsturgeon** [comment]
There's a few approaches to avoid duplication: 

1. [jest.each](https://blog.theodo.com/2018/09/jest-each-tests/)
2. defining common schema in a context (outside of it()) and just tweaking it a bit for each one, instead of redefining the whole thing
3. make some "fixture files"

I'm usually quite content with 1 or 2 but 3 can lead to awkward changes when multiple people do different things with the fixtures. 

Also keep in mind we don't need to fix this now, if we keep on boy scouting as we go it'll get better over time.

**Comment by nulltoken** [comment]
@P0lip My thoughts were around a combination of @philsturgeon points 1 & 2. However, I'm also fine with the `we don't need to fix this now, if we keep on boy scouting as we go it'll get better over time` way. You've done a tremendous job and maybe refactoring the examples would just deserve a neat issue to keep track of it

**Comment by P0lip** [comment]
@nulltoken @philsturgeon alright, did some deduplication.
We could probably do a little bit more here and there, but it should be already much better

**Review by philsturgeon — APPROVED** [review]

**Pull request merged**